bufferevent_sock.c

```diff
@@ -463,11 +462,14 @@ bufferevent_connect_getaddrinfo_cb(int result, struct evutil_addrinfo *ai,
 	struct bufferevent_private *bev_p =
 	    EVUTIL_UPCAST(bev, struct bufferevent_private, bev);
 	int r;
+	int option;
 	BEV_LOCK(bev);
 
 	bufferevent_unsuspend_write_(bev, BEV_SUSPEND_LOOKUP);
 	bufferevent_unsuspend_read_(bev, BEV_SUSPEND_LOOKUP);
 
+	/* if this callback is in bufferevent_socket_connect_hostname() then defer event cb */
+	option = (bev_p->dns_request == NULL ? BEV_OPT_DEFER_CALLBACKS : 0);
```

Comment:
In general a good thing to do, but this changes the behavior of the API without `dns_base`, not a good thing to do, let me think on this a little

Reply:
@azat consider the following cases
## 1. simple case
```c
static void eventcb(struct bufferevent *bev, short what, void *ctx)
{
        socks5_conn *conn = (socks5_conn*)ctx;
        if (what & (BEV_EVENT_EOF|BEV_EVENT_ERROR)) {
                // notify error
                conn->cb(NULL, conn->arg);

                free(conn);
                bufferevent_free(bev);
        } else if (what & BEV_EVENT_CONNECTED) {
                const char headers[] =
                        "GET / HTTP/1.1\r\n"
                        "Connection: Keep-Alive\r\n"
                        "\r\n";
                bufferevent_write(bev, headers, sizeof(headers) - 1); // without ending '\0'
        }
}

socks5_conn* connect_socks5(struct event_base *evbase, struct evdns_base *evdns_base, const char *hostname, int port, connect_callback cb, void *arg)
{
        struct bufferevent *bev = NULL;
        socks5_conn *conn = (socks5_conn*)calloc(1, sizeof(socks5_conn));
        if (NULL == conn) {
                goto fail;
        }

        /* do somthing with conn */
        conn->cb = cb;

        bev = bufferevent_socket_new(evbase, -1, BEV_OPT_CLOSE_ON_FREE/*|BEV_OPT_DEFER_CALLBACKS*/);
        if (NULL == bev) {
                LOGE("bufferevent create failed");
                goto fail;
        }

        conn->client = bev;
        /* must be set before bufferevent_socket_connect_hostname(), otherwise no chance to catch error */
        bufferevent_setcb(bev, readcb, NULL, eventcb, conn);

        if (bufferevent_socket_connect_hostname(bev, evdns_base, PF_UNSPEC, hostname, port) < 0) {
                LOGE("bufferevent connect %s:%d failed", hostname, port);
                goto fail;
        }

        // Because eventcb() might have been called when name resolve can be done immediately, so we cannot do anything with bev or conn here.
        // If you want, you must use BEV_OPT_DEFER_CALLBACKS to force eventcb() executed after bufferevent_socket_connect_hostname().

        return conn; /* not safe, unless defered event cb */

fail:
        if (bev)
                bufferevent_free(bev);
        if (conn)
                free(conn);
        return NULL
}

```
If we want to return conn object in connect_socks5(), we must defer eventcb() to ensure conn object not be freed early. 

## 2. more complex case
We use SSL connection instead of simple TCP connection. 

In this case, we need a sock_eventcb() for bufferevent_socket_connect_hostname(), and a noraml ssl_eventcb(). But this case need two event callback is too complex, and need extra document to teach users.
```c
        bev = bufferevent_socket_new(evbase, -1, BEV_OPT_CLOSE_ON_FREE/*|BEV_OPT_DEFER_CALLBACKS*/);
        if (NULL == bev) {
                LOGE("bufferevent create failed");
                goto fail;
        }

        conn->client = bev;
        /* must be set before bufferevent_socket_connect_hostname() */
        bufferevent_setcb(bev, readcb, NULL, sock_eventcb, conn);

        if (bufferevent_socket_connect_hostname(bev, evdns_base, PF_UNSPEC, hostname, port) < 0) {
                LOGE("bufferevent connect %s:%d failed", hostname, port);
                goto fail;
        }

        SSL *ssl = SSL_new(ssl_ctx);
        struct bufferevent *b_ssl = bufferevent_openssl_filter_new(base, bev, ssl, BUFFEREVENT_SSL_CONNECTING,
                        BEV_OPT_CLOSE_ON_FREE|BEV_OPT_DEFER_CALLBACKS);
        if (!b_ssl) {
                goto fail;
        }
        /* set normal event cb */
        bufferevent_setcb(bev, readcb, NULL, ssl_eventcb, conn);
```



## To hundle the problem most essentially
* The evdns_getaddrinfo() must not call evdns_getaddrinfo_cb()  directly in itself, but the fix is too complex.
* Or, bufferevent_socket_connect_hostname() always defer its name resolve function evutil_getaddrinfo_async() . moderate complex.

Reply:
Yeah you are right, I see your point. Maybe we should update documentation at least for this, not sure that calling eventcb always deferred there is such a great idea.
Or maybe we should introduce another flag like `BEV_OPT_SAFE` (or something similar)
Or maybe I'm just making noise, and there is nothing critical in calling eventcb always deferred.

Thoughts?